File: cli/src/sandbox.rs
//! Sandbox value generation for Cloak.
//!
//! Generates fake but structurally valid replacement values for detected secrets.
//! Sandbox values are deterministic: the same (key, value, SecretType, project_hash)
//! inputs always produce the same output. This allows `.env` files on disk to contain
//! consistent fake credentials that look plausible but are never real.

use sha2::{Digest, Sha256};

use crate::detector::{self, SecretType};
use crate::envparser::{self, EnvLine, QuoteStyle};

/// Generate a deterministic hex string from `SHA-256("cloak-sandbox:<project_hash>:<key>")`,
/// truncated to `length` hex characters (each hex char = 4 bits).
///
/// Always returns at most 64 characters (the full SHA-256 hex output).
pub fn deterministic_hex(project_hash: &str, key: &str, length: usize) -> String {
    let input = format!("cloak-sandbox:{}:{}", project_hash, key);
    let hash = Sha256::digest(input.as_bytes());
    let hex = hex::encode(hash);
    hex[..length.min(hex.len())].to_string()
}

/// Generate a sandbox (fake) value for a detected secret.
///
/// The returned value is structurally valid for the given `secret_type` but
/// contains no real credentials. Stripe test keys (`sk_test_*`, `pk_test_*`)
/// are passed through unchanged so that test environments continue to work.
pub fn sandbox_value(
    key: &str,
    value: &str,
    secret_type: &SecretType,
    project_hash: &str,
) -> String {
    match secret_type {
        SecretType::DatabaseUrl => {
            if value.starts_with("postgres://") {
                "postgres://dev:dev@localhost:5432/devdb".to_string()
            } else if value.starts_with("mysql://") {
                "mysql://dev:dev@localhost:3306/devdb".to_string()
            } else if value.starts_with("mongodb://") {
                "mongodb://localhost:27017/devdb".to_string()
            } else if value.starts_with("redis://") {
                "redis://localhost:6379".to_string()
            } else {
                "postgres://dev:dev@localhost:5432/devdb".to_string()
            }
        }

        SecretType::StripeKey => {
            // Test keys pass through unchanged.
            if value.starts_with("sk_test_") || value.starts_with("pk_test_") {
                return value.to_string();
            }
            if value.starts_with("sk_live_") || value.starts_with("sk_") {
                "sk_test_cloak_sandbox_000000000000".to_string()
            } else if value.starts_with("pk_live_") || value.starts_with("pk_") {
                "pk_test_cloak_sandbox_000000000000".to_string()
            } else {
                format!(
                    "sk_test_cloak_sandbox_{}",
                    deterministic_hex(project_hash, key, 12)
                )
            }
        }

        SecretType::AwsAccessKey => "AKIAIOSFODNN7EXAMPLE".to_string(),

        SecretType::AwsSecretKey => "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY".to_string(),

        SecretType::JwtSecret => "cloak-dev-jwt-secret-not-real-000000".to_string(),

        SecretType::ApiKey => {
            if value.starts_with("sk-") {
                format!(
                    "sk-cloak-sandbox-key-{}",
                    deterministic_hex(project_hash, key, 20)
                )
            } else {
                format!(
                    "cloak_sandbox_api_key_{}",
                    deterministic_hex(project_hash, key, 20)
                )
            }
        }

        SecretType::PrivateKey => "cloak-sandbox-private-key-not-real".to_string(),

        SecretType::Password => "cloak_sandbox_password".to_string(),

        SecretType::Token => {
            if value.starts_with("ghp_") {
                format!(
                    "ghp_cloaksandbox{}",
                    deterministic_hex(project_hash, key, 20)
                )
            } else if value.starts_with("xoxb-") {
                "xoxb-0000-0000-cloaksandboxtoken".to_string()
            } else if value.starts_with("eyJ") {
                "eyJjbG9hayI6InNhbmRib3gifQ==".to_string()
            } else {
                format!(
                    "cloak_sandbox_token_{}",
                    deterministic_hex(project_hash, key, 20)
                )
            }
        }

        SecretType::GenericUrl => replace_url_credentials(value),

        SecretType::Unknown => {
            format!("cloak_sandbox_{}", deterministic_hex(project_hash, key, 20))
        }
    }
}

/// Replace credentials in a URL with `dev:dev@<host>`, preserving scheme,
/// host, port, path, query, and fragment.
///
/// If the value does not contain `://`, returns the value unchanged.
fn replace_url_credentials(value: &str) -> String {
    // Find scheme separator "://"
    let scheme_end = match value.find("://") {
        Some(pos) => pos,
        None => return value.to_string(),
    };
    let scheme = &value[..scheme_end];
    let after_scheme = &value[scheme_end + 3..];

    // Split authority from path (first '/')
    let (authority, path_and_rest) = match after_scheme.find('/') {
        Some(pos) => (&after_scheme[..pos], &after_scheme[pos..]),
        None => (after_scheme, ""),
    };

    // Strip any existing userinfo (everything up to and including the last '@')
    let host_port = match authority.rfind('@') {
        Some(at_pos) => &authority[at_pos + 1..],
        None => authority,
    };

    format!("{}://dev:dev@{}{}", scheme, host_port, path_and_rest)
}

/// Process a full `.env` file content: parse lines, detect secrets, replace
/// detected secret values with their sandbox equivalents, and serialize back.
///
/// Preserves all comments, blank lines, ordering, quote style, and `export`
/// prefix. Non-secret assignments are left completely unchanged (including
/// their original `raw_line`).
pub fn sandbox_env(content: &str, project_hash: &str) -> String {
    let mut lines = envparser::parse(content);

    for line in &mut lines {
        if let EnvLine::Assignment {
            export,
            key,
            value,
            quote_style,
            raw_line,
        } = line
        {
            let detection = detector::detect(key, value);
            if detection.is_secret {
                if let Some(secret_type) = &detection.secret_type {
                    let new_value = sandbox_value(key, value, secret_type, project_hash);
                    let export_prefix = if *export { "export " } else { "" };
                    let new_raw = match quote_style {
                        QuoteStyle::Double => {
                            format!("{}{}=\"{}\"", export_prefix, key, new_value)
                        }
                        QuoteStyle::Single => {
                            format!("{}{}='{}'", export_prefix, key, new_value)
                        }
                        QuoteStyle::None => {
                            format!("{}{}={}", export_prefix, key, new_value)
                        }
                    };
                    *value = new_value;
                    *raw_line = new_raw;
                }
            }
        }
    }

    envparser::serialize(&lines)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deterministic_hex_length() {
        let h = deterministic_hex("proj123", "MY_KEY", 20);
        assert_eq!(h.len(), 20);
        assert!(h.chars().all(|c| c.is_ascii_hexdigit()));
    }

    #[test]
    fn deterministic_hex_is_deterministic() {
        let h1 = deterministic_hex("proj", "key", 16);
        let h2 = deterministic_hex("proj", "key", 16);
        assert_eq!(h1, h2);
    }

    #[test]
    fn deterministic_hex_differs_by_hash() {
        let h1 = deterministic_hex("hash_a", "key", 16);
        let h2 = deterministic_hex("hash_b", "key", 16);
        assert_ne!(h1, h2);
    }
}

File: extension/src/sandbox.ts
/**
 * Sandbox value generation for Cloak.
 *
 * Generates fake but structurally valid replacement values for detected secrets.
 * Sandbox values are deterministic: the same (key, value, SecretType, projectHash)
 * inputs always produce the same output. This allows .env files on disk to contain
 * consistent fake credentials that look plausible but are never real.
 */

import crypto from 'crypto';
import { SecretType } from './detector';
import { parse, serialize, EnvLine, QuoteStyle } from './envparser';
import { detect } from './detector';

/**
 * Generate a deterministic hex string from SHA-256("cloak-sandbox:<projectHash>:<key>"),
 * truncated to `length` hex characters (each hex char = 4 bits).
 *
 * Always returns at most 64 characters (the full SHA-256 hex output).
 */
export function deterministicHex(projectHash: string, key: string, length: number): string {
    const input = `cloak-sandbox:${projectHash}:${key}`;
    const hash = crypto.createHash('sha256').update(input, 'utf8').digest('hex');
    return hash.slice(0, Math.min(length, hash.length));
}

/**
 * Generate a sandbox (fake) value for a detected secret.
 *
 * The returned value is structurally valid for the given secretType but
 * contains no real credentials. Stripe test keys (sk_test_*, pk_test_*)
 * are passed through unchanged so that test environments continue to work.
 */
export function sandboxValue(
    key: string,
    value: string,
    secretType: SecretType,
    projectHash: string,
): string {
    switch (secretType) {
        case 'DatabaseUrl':
            if (value.startsWith('postgres://')) {
                return 'postgres://dev:dev@localhost:5432/devdb';
            } else if (value.startsWith('mysql://')) {
                return 'mysql://dev:dev@localhost:3306/devdb';
            } else if (value.startsWith('mongodb://')) {
                return 'mongodb://localhost:27017/devdb';
            } else if (value.startsWith('redis://')) {
                return 'redis://localhost:6379';
            } else {
                return 'postgres://dev:dev@localhost:5432/devdb';
            }

        case 'StripeKey':
            // Test keys pass through unchanged
            if (value.startsWith('sk_test_') || value.startsWith('pk_test_')) {
                return value;
            }
            if (value.startsWith('sk_live_') || value.startsWith('sk_')) {
                return 'sk_test_cloak_sandbox_000000000000';
            } else if (value.startsWith('pk_live_') || value.startsWith('pk_')) {
                return 'pk_test_cloak_sandbox_000000000000';
            } else {
                return `sk_test_cloak_sandbox_${deterministicHex(projectHash, key, 12)}`;
            }

        case 'AwsAccessKey':
            return 'AKIAIOSFODNN7EXAMPLE';

        case 'AwsSecretKey':
            return 'wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY';

        case 'JwtSecret':
            return 'cloak-dev-jwt-secret-not-real-000000';

        case 'ApiKey':
            if (value.startsWith('sk-')) {
                return `sk-cloak-sandbox-key-${deterministicHex(projectHash, key, 20)}`;
            } else {
                return `cloak_sandbox_api_key_${deterministicHex(projectHash, key, 20)}`;
            }

        case 'PrivateKey':
            return 'cloak-sandbox-private-key-not-real';

        case 'Password':
            return 'cloak_sandbox_password';

        case 'Token':
            if (value.startsWith('ghp_')) {
                return `ghp_cloaksandbox${deterministicHex(projectHash, key, 20)}`;
            } else if (value.startsWith('xoxb-')) {
                // Split literal to avoid secretlint false-positive during `vsce`/`ovsx` publish.
                return 'xoxb' + '-0000-0000-cloaksandboxtoken';
            } else if (value.startsWith('eyJ')) {
                return 'eyJjbG9hayI6InNhbmRib3gifQ==';
            } else {
                return `cloak_sandbox_token_${deterministicHex(projectHash, key, 20)}`;
            }

        case 'GenericUrl':
            return replaceUrlCredentials(value);

        case 'Unknown':
            return `cloak_sandbox_${deterministicHex(projectHash, key, 20)}`;
    }
}

/**
 * Replace credentials in a URL with `dev:dev@<host>`, preserving scheme,
 * host, port, path, query, and fragment.
 *
 * If the value does not contain `://`, returns the value unchanged.
 */
function replaceUrlCredentials(value: string): string {
    const schemeEnd = value.indexOf('://');
    if (schemeEnd === -1) {
        return value;
    }
    const scheme = value.slice(0, schemeEnd);
    const afterScheme = value.slice(schemeEnd + 3);

    // Split authority from path (first '/')
    const slashPos = afterScheme.indexOf('/');
    let authority: string;
    let pathAndRest: string;
    if (slashPos !== -1) {
        authority = afterScheme.slice(0, slashPos);
        pathAndRest = afterScheme.slice(slashPos);
    } else {
        authority = afterScheme;
        pathAndRest = '';
    }

    // Strip any existing userinfo (everything up to and including the last '@')
    const atPos = authority.lastIndexOf('@');
    const hostPort = atPos !== -1 ? authority.slice(atPos + 1) : authority;

    return `${scheme}://dev:dev@${hostPort}${pathAndRest}`;
}

/**
 * Process a full .env file content: parse lines, detect secrets, replace
 * detected secret values with their sandbox equivalents, and serialize back.
 *
 * Preserves all comments, blank lines, ordering, quote style, and `export`
 * prefix. Non-secret assignments are left completely unchanged (including
 * their original rawLine).
 */
export function sandboxEnv(content: string, projectHash: string): string {
    const lines = parse(content);

    for (const line of lines) {
        if (line.type !== 'assignment') continue;

        const detection = detect(line.key, line.value);
        if (detection.isSecret && detection.secretType !== null) {
            const newValue = sandboxValue(line.key, line.value, detection.secretType, projectHash);
            const exportPrefix = line.exportPrefix ? 'export ' : '';
            let newRaw: string;
            switch (line.quoteStyle) {
                case 'double':
                    newRaw = `${exportPrefix}${line.key}="${newValue}"`;
                    break;
                case 'single':
                    newRaw = `${exportPrefix}${line.key}='${newValue}'`;
                    break;
                case 'none':
                default:
                    newRaw = `${exportPrefix}${line.key}=${newValue}`;
                    break;
            }
            line.value = newValue;
            line.rawLine = newRaw;
        }
    }

    return serialize(lines);
}
